lightningd/bitcoind.c

```diff
@@ -102,98 +102,75 @@ static void bitcoin_plugin_error(struct bitcoind *bitcoind, const char *buf,
 	      toks->end - toks->start, buf + toks->start);
 }
 
-/* `getfeerate`
+/* `estimatefees`
  *
  * Gather feerate from our Bitcoin backend. Will set the feerate to `null`
  * if estimation failed.
  *
  * Plugin response:
  * {
- *	"feerate": <fee per KB>,
+ *	"urgent": <btc per KB>,
+ *	"normal": <btc per KB>,
+ *	"slow": <btc per KB>
  * }
  */
 
 struct estimatefee_call {
 	struct bitcoind *bitcoind;
-	size_t i;
-	const u32 *blocks;
-	const char **estmode;
-
 	void (*cb)(struct bitcoind *bitcoind, const u32 satoshi_per_kw[],
 		   void *);
 	void *arg;
-	u32 *satoshi_per_kw;
 };
 
-static void do_one_estimatefee(struct bitcoind *bitcoind,
-			       struct estimatefee_call *call);
-
-static void getfeerate_callback(const char *buf, const jsmntok_t *toks,
-				const jsmntok_t *idtok,
-				struct estimatefee_call *call)
+static void estimatefees_callback(const char *buf, const jsmntok_t *toks,
+				  const jsmntok_t *idtok,
+				  struct estimatefee_call *call)
 {
 	const jsmntok_t *resulttok, *feeratetok;
-	u64 feerate;
+	const char *targets[] = {"urgent", "normal", "slow"};
+	u32 *feerates = tal_arr(call, u32, 3);
```

Comment:
> can you just straight declare this as an array of size 3? not sure how this interacts with the cb below. it seems like it'd be fine since you clean up call at the end of this routine anyway.

I'm almost sure that it has to be allocated on the heap to be passed to the cb.

> if you use a constant for the size, you can re-use it for the for loop count below (instead of tal_count)
e.g. #define FEERATE_COUNT 3; u32 feerates[FEERATE_COUNT];

Ok, I'll use chaintopology's one then.

Reply:
Done